File: 0392. Is Subsequence/dp.java
class Solution {
    public boolean isSubsequence(String s, String t) {
        if(s.length() == 0)
            return true;
        int len1 = s.length(), len2 = t.length();
        boolean[][] dp = new boolean[len1 + 1][len2 + 1];
        for(int i = 0; i < len1 + 1; ++i){
            for(int j = 0; j < len2 + 1; ++j){
                dp[i][j] = false;
            }
        }
        for(int i = 0; i < len2 + 1; ++i){
            dp[0][i] = true;
        }
        
        for(int i = 1; i < len1 + 1; ++i){
            for(int j = 1; j < len2 + 1; ++j){
                if(s.charAt(i - 1) == t.charAt(j - 1)){
                    dp[i][j] = dp[i - 1][j - 1];
                }else{
                    dp[i][j] = dp[i - 1][j] && dp[i][j - 1];
                }
            }
        }
        
        return dp[len1][len2];
    }
}

File: 0010. Regular Expression Matching/dp.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int len1 = s.size(), len2 = p.size();
        vector<vector<bool>> dp(len1 + 1, vector<bool>(len2 + 1, false));
        dp[0][0] = true;
        for(int i = 1; i < len2 + 1; ++i){
            // if(p[i - 1] == '*'){
            //     dp[0][i] = dp[0][i - 1];
            // }else{
            //     break;
            // }
            if(p[i - 1] == '*' && dp[0][i - 2])
                dp[0][i] = true;
        }
        
        for(int i = 1; i < len1 + 1; ++i){
            for(int j = 1; j < len2 + 1; ++j){
                if(p[j - 1] == s[i - 1] || p[j - 1] == '.'){
                    dp[i][j] = dp[i - 1][j - 1];
                }else if(p[j - 1] == '*'){
                    if(p[j - 2] != s[i - 1] && p[j - 2] != '.'){
                        dp[i][j] = dp[i][j - 2];
                    }else{
                        dp[i][j] = (dp[i][j - 2] || dp[i][j - 1] || dp[i - 1][j]);
                    }
                }
            }
        }
        
        return dp[len1][len2];
    }
};